File: src/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def build(self, prel, prer, inl, inr):
        root = self.preorder[prel]
        index = self.inorder.index(root, inl, inr)
        root_node = TreeNode(root)
        if index - inl:
            root_node.left = self.build(
                prel + 1, prel + index - inl + 1,
                inl, index
            )
        if inr - index - 1:
            root_node.right = self.build(
                prel + index - inl + 1, prer,
                index + 1, inr
            )
        return root_node

    # @param {integer[]} preorder
    # @param {integer[]} inorder
    # @return {TreeNode}
    def buildTree(self, preorder, inorder):
        if preorder == [] and inorder == []:
            return None
        self.length = len(preorder)
        self.preorder, self.inorder = preorder, inorder
        return self.build(
            0, self.length,
            0, self.length
        )

if __name__ == "__main__":
    solution = Solution()
    node = solution.buildTree([1, 2, 3], [2, 1, 3])
    print node.val, node.left.val, node.right.val

File: src/algorithms/cpp/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    vector <int> preorder, inorder;

private:
    TreeNode* build(int prel, int prer, int inl, int inr) {
        int root = this->preorder[prel];
        int index = -1;
        for (int i = inl; i < inr; ++ i)
            if (this->inorder[i] == root) {
                index = i;
                break;
            }
        TreeNode *root_node = new TreeNode(root);
        if (index - inl > 0)
            root_node->left = build(prel + 1, prel + index - inl + 1, inl, index);
        if (inr - index - 1 > 0)
            root_node->right = build(prel + index - inl + 1, prer, index + 1, inr);
        return root_node;
    }

public:
    TreeNode* buildTree(vector <int> &preorder, vector <int> &inorder) {
        this->preorder = preorder, this->inorder = inorder;
        if (SZ(preorder) == 0)      return NULL;
        return build(0, SZ(preorder), 0, SZ(inorder));
    }
};

int main() {
    return 0;
}
